File: 20-ValidParentheses/20-ValidParentheses.java
// Last updated: 5/30/2025, 7:35:13 PM
class Solution {
    public boolean isValid(String s) {
        HashMap<Character,Character> map = new HashMap<Character,Character>();
        map.put('[',']');
        map.put('(',')');
        map.put('{','}');
        Stack<Character> stack = new Stack<Character>();
        for (char c : s.toCharArray()){
            if (map.containsKey(c)) {
                stack.push(c);
            }
            else if (stack.isEmpty() || map.get(stack.pop())!=c) {
                return false;
            }
        }
        return stack.isEmpty();
    }
}

File: 20-ValidParentheses/20-ValidParentheses.py
# Last updated: 6/11/2025, 2:58:39 PM
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        closeToOpen = {')':'(', ']':'[', '}':'{'}
        for c in s:
            if c in closeToOpen:
                if stack and stack[-1] == closeToOpen[c]:
                    stack.pop()
                else:
                    return False
            else:
                stack.append(c)
        return len(stack) == 0